merlin-server/sql/merlin/plan_collaboration.sql

```diff
@@ -0,0 +1,188 @@
+  source_scheduling_goal_id integer,
+  created_at timestamptz not null,
+  last_modified_at timestamptz not null,
+  start_offset interval not null,
+  type text not null,
+  arguments merlin_argument_set not null,
+  last_modified_arguments_at timestamptz not null,
+  metadata merlin_activity_directive_metadata_set
+);
+
+-- Captures the state of a plan and all of its activities
+create or replace function create_snapshot(plan_id integer)
+  returns integer -- snapshot id inserted into the table
+  language plpgsql as $$
+  declare
+    validate_planid integer;
+    inserted_snapshot_id integer;
+begin
+  select id from plan where plan.id = plan_id into validate_planid;
+  if validate_planid is null then
+    raise exception 'Plan % does not exist.', plan_id;
+  end if;
+
+  insert into plan_snapshot(plan_id, revision, name, duration, start_time)
+    select id, revision, name, duration, start_time
+    from plan where id = plan_id
+    returning snapshot_id into inserted_snapshot_id;
+  insert into plan_snapshot_activities(snapshot_id, id, name, tags, source_scheduling_goal_id, created_at, last_modified_at, start_offset, type, arguments, last_modified_arguments_at, metadata)
+    select
+      inserted_snapshot_id,                                   --this is the snapshot id
+      id, name, tags,source_scheduling_goal_id, created_at,   -- these are the rest of the data for an activity row
+      last_modified_at, start_offset, type, arguments, last_modified_arguments_at, metadata
+    from activity_directive where activity_directive.plan_id = create_snapshot.plan_id;
+
+  --all snapshots in plan_latest_snapshot for plan plan_id become the parent of the current snapshot
+  insert into plan_snapshot_parent(snapshot_id, parent_snapshot_id)
+    select inserted_snapshot_id, snapshot_id
+    from plan_latest_snapshot where plan_latest_snapshot.plan_id = create_snapshot.plan_id;
+
+  --remove all of those entries from plan_latest_snapshot and add this new snapshot.
+  delete from plan_latest_snapshot where plan_latest_snapshot.plan_id = create_snapshot.plan_id;
+  insert into plan_latest_snapshot(plan_id, snapshot_id) values (create_snapshot.plan_id, inserted_snapshot_id);
+
+  return inserted_snapshot_id;
+  end;
+$$;
+
+/*
+  Copies all of a given plan's properties and activities into a new plan with the specified name.
+  When duplicating a plan, a snapshot is created of the original plan.
+  Additionally, that snapshot becomes the latest snapshot of the new plan.
+*/
+  create or replace function duplicate_plan(plan_id integer, new_plan_name text)
+  returns integer -- plan_id of the new plan
+  security definer
+  language plpgsql as $$
+  declare
+    validate_plan_id integer;
+    new_plan_id integer;
+    created_snapshot_id integer;
+begin
+  select id from plan where plan.id = duplicate_plan.plan_id into validate_plan_id;
+  if(validate_plan_id is null) then
+    raise exception 'Plan % does not exist.', plan_id;
+  end if;
```

Comment:
this should be covered by key constraints

Reply:
Assuming that the key constraint you're referring to is the one you proposed on the plan table, since we work with another table before attempting the insert, I disagree with removing this check.